### PR TITLE
Support for ThreadX/AzureRTOS

### PR DESCRIPTION
**Pull request by newbrain**
Here is, as promised (threatened?) the PR for a ThreadX/AzureRTOS plug-in.

### General info
- The plug-in is named threadx.
- The code is, as one can expect, heavily modelled on the existing RTOS plug-ins.
- It has been autopep8'ed
- I'm a C programmer, mostly, this might show here and there.
- The plug-in has been tested with a NUCLEO-STM32F411RE and an OKDO-E1 (LPC55S69) and some test programs, using also hard FP but not including use of the secure features on LPC55S69.
- Tests have been done both with command line debugging and from inside VS Code.
- The stack frames look correct (registers and variables as expected), but they differ from ThreadX docs, see azure-rtos/threadx#70
- I have noticed that if I leave pyocd gdbserver running, quit gdb and then start a new gdb session, the plug-in is correctly loaded but threads are disabled in the new session until I issue a monitor thread enable. After that, the expected behaviour resumes.

### Running the internal test suite
- All unit tests on Windows 10 Pro 20H2 and Linux Mint 20.1 passes, with the exception of #1050 on Windows.
- Automatic tests always fail on Windows, with all the following boards:
  - NUCLEO STM32F411RE 
  - NUCLEO STM32F401RE 
  - NUCLEO STM32L476RG 
  - OKDO E1
  - Some others I've forgotten...
  - NXP MIMXRT1020-EVK
- The same happens on Linux, with the exception of the NUCLEO STM32L476RG, which passes all test.
- The failing tests (both systems) are  usually "Gdb Test" and "Flash Loader Test". Not 100% repeatable.
- MIMXRT1020-EVK behaves differently, see below:
```
------ TEST RESULTS ------
Target         Test                 Result         Time

mimxrt1020     Basic Test           Fail           2.302
mimxrt1020     Json Lists Test      Pass           10.825
mimxrt1020     Connect Test         Pass           9.743
mimxrt1020     Speed Test           Fail           2.163
mimxrt1020     Cortex Test          Fail           12.349
mimxrt1020     Concurrency Test     Fail           2.170
mimxrt1020     Flash Test           Pass           2718.288
mimxrt1020     Flash Loader Test    Pass           18.418
mimxrt1020     Debug Context Test   Pass           1.898
mimxrt1020     Gdb Test             Pass           69.294
mimxrt1020     Commands Test        Pass           8.540
mimxrt1020     Commander Test       Pass           0.852
```

This should not have anything to do with the plug-in itself, though, as a clean master version behaves the same.

### A possible improvement
To generate the threads' list, the plug-in walks the (doubly) linked list of created threads (so it's able to report also completed and terminated threads - of course not deleted ones).
Walking this list, and reading name, state and priority attributes of the thread relies on offsets inside the TX_THREAD structure.
Unfortunately, if the user has defined some thread extensions (using TX_THREAD_EXTENSION_0..3) these offsets will be wrong.

https://github.com/azure-rtos/threadx/blob/32e3b3b25f4b17ff51ce8a6bad8d9eb32729a0f6/common/inc/tx_api.h#L410-L418

To have some degree of confidence, the plug-in checks that each thread structure reached has the right id ('THRD' chars at offset 0)  and that the total number of threads matches the created threads count (yes, there's a small window in thread creation and deletion  where this will fail, maybe a + or - 1 tolerance should be added).

The ideal solution would be to read the offsets from the .elf, I don't know whether that is possible.
A thread is not guaranteed to exist, as ThreadX uses no Idle thread and the timer thread can be done away with and tucked in an ISR.

### CODE REVIEW

**Comment by flit** [comment]
> I have noticed that if I leave pyocd gdbserver running, quit gdb and then start a new gdb session, the plug-in is correctly loaded but threads are disabled in the new session until I issue a monitor thread enable. After that, the expected behaviour resumes.

The gdbserver includes some logic to disable reporting threads until the first resume after a connect or reset. This is to work around the problem of stale thread data after you load updated code but RAM still contains partially valid RTOS data structures. `monitor thread enable` simply overrides this logic. (Jeez I need to document all this…)

> Automatic tests always fail on Windows

I'm not too worried. Running the functional tests on STM32 and i.MX RT boards has always been problematic (though will have to be addressed in the new CI system!) And of course the functional tests don't test RTOS plugins (yet).

> A thread is not guaranteed to exist…

Is a "live" fallback thread created? Or put another way, what does the user see when there is no thread and they halt in gdb?

**Comment by newbrain** [comment]
Thanks for the explanation, it makes sense.

> Is a "live" fallback thread created? Or put another way, what does the user see when there is no thread and they halt in gdb?

For the case we are still in initialization phase:
ThreadX has an internal variable (_tx_thread_system_state) that tells you whether:
* It's initializing (in two steps).
* It's in the ready state (0).
* On some ports, if It's inside a ThreadX interrupt handler (for cortex-m it's always 0).

In the plug-in, this is checked by the is_enabled @property in ThreadXThreadProvider.
In gdb, it appears as a single non-RTOS thread.
When inside an exception handler, an extra (fictitious) thread is reported.

Not having any threads is a bit of a corner case, but it is indeed possible:
* Have the timer task served in IRS (it's compile time #define) 
* Do not create any thread, and register a timeout during the initialization phase, with threads creation in its callback function.
* Until the timer expires, there'll be 0 threads, the code will spin in a "wait here" loop.

And you probably got me here - I did not test this scenario so I cannot answer.
I will do ASAP (need to put together a test program), and in case add a check for 0 threads with a default action.

**Comment by flit** [comment]
Since it takes a bit of an unrealistic scenario to have no threads, I'm not really worried about it at this point. Thanks for the explanation of the required sequence.

What about when there are no running or ready threads? If I remember correctly, ThreadX stays in a loop in the scheduler until a thread becomes ready. But it's been _years_ since I used ThreadX, so I guess it may have changed.

**Comment by newbrain** [comment]
Oh, for me it's a new thing: I noticed Microsoft acquired it, so I was curious.
I don't dislike it, the naming is less confusing than FreeRTOS, it does not have FreeRTOS's limitation of special APIs for IRS (and the one on IRQ priority).

A thread in ThreadX can be in a number of states, they will be reported in gdb along the other thread info.
See below:
```
(gdb) tb main
Temporary breakpoint 1 at 0x800041c: file Src/sample_threadx.c, line 62.
(gdb) c
Continuing.
Note: automatically using hardware breakpoints for read-only addresses.

Temporary breakpoint 1, main () at Src/sample_threadx.c:62
62          tx_kernel_enter();
(gdb) info threads
  Id   Target Id         Frame
* 1    Thread 1          main () at Src/sample_threadx.c:62
(gdb) c
Continuing.
[New Thread 536871640]
[New Thread 536882300]
[New Thread 536881408]
[New Thread 536871044]
[New Thread 536872108]
[New Thread 536871456]
[New Thread 536881748]
[New Thread 536871820]
[New Thread 536871264]

Thread 2 "thread 1" received signal SIGINT, Interrupt.
[Switching to Thread 536871640]
0x08003b48 in _tx_queue_send (queue_ptr=0x200029b4 <queue_0>, source_ptr=0x200029f0 <thread_1_messages_sent>,
    wait_option=4294967295) at ThreadX/common/src/tx_queue_send.c:147
147                 TX_QUEUE_MESSAGE_COPY(source, destination, size)
(gdb) info threads
  Id   Target Id                                                      Frame
* 2    Thread 536871640 "thread 1" (Ready; Priority 16)               0x08003b48 in _tx_queue_send (
    queue_ptr=0x200029b4 <queue_0>, source_ptr=0x200029f0 <thread_1_messages_sent>, wait_option=4294967295)
    at ThreadX/common/src/tx_queue_send.c:147
  3    Thread 536882300 "System Timer Thread" (Suspended; Priority 0) 0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  4    Thread 536881408 "thread 0" (Sleep; Priority 1)                0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  5    Thread 536871044 "thread 2" (Ready; Priority 16)               0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  6    Thread 536872108 "thread 3" (Semaphore; Priority 8)            0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  7    Thread 536871456 "thread 4" (Sleep; Priority 8)                0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  8    Thread 536881748 "thread 5" (EventFlag; Priority 4)            0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  9    Thread 536871820 "thread 6" (Mutex; Priority 8)                0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  10   Thread 536871264 "thread 7" (Sleep; Priority 8)                0x08004a12 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
```
So, "thread 1" (2) was the current one, interrupted and is in _Ready_ state, the "System Timer Thread" is _Suspended_ (waiting for a service) and others are waiting for _EventFlags_, _Sleep_ing etc.
After all, with only one core, only one thread is really executing at any time.

As a quick test for the zero threads issue:
I patched the plug-in to always give an empty thread list, and the result is that the current position is shown, it is possible to step etc. BUT **warning: Remote failure reply: E01** is reported:
```
(gdb) c
Continuing.
warning: Remote failure reply: E01

Program stopped.
0x080007e0 in thread_2_entry (thread_input=2) at Src/sample_threadx.c:248
248             if ((status != TX_SUCCESS) || (received_message != thread_2_messages_received))
(gdb) bt
#0  0x080007e0 in thread_2_entry (thread_input=2) at Src/sample_threadx.c:248
#1  0x08004392 in _tx_thread_shell_entry () at ThreadX/common/src/tx_thread_shell_entry.c:114
#2  0xfffffffe in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
[...]
(gdb) n
warning: Remote failure reply: E01

Program stopped.
_txe_queue_receive (queue_ptr=0x200029b4 <queue_0>, destination_ptr=0x20001138 <memory_area+3032>,
    wait_option=4294967295) at ThreadX/common/src/txe_queue_receive.c:102
102         else if (queue_ptr -> tx_queue_id != TX_QUEUE_ID)
(gdb) info threads
  Id   Target Id         Frame
* 1    Thread 1          _txe_queue_receive (queue_ptr=0x200029b4 <queue_0>,
    destination_ptr=0x20001138 <memory_area+3032>, wait_option=4294967295)
    at ThreadX/common/src/txe_queue_receive.c:102
```

**Comment by flit** [comment]
I used ThreadX extensively for nearly a decade, though never on Cortex-M, and it's one of my favourite RTOSes. In some ways, I based my [Argon](/flit/argos-rtos) RTOS on it. I was happy to see Microsoft buying and open sourcing it.

The thread status reporting looks great, just what I would expect.

Really, don't worry about the no threads issue right now. You can add that in a follow-up PR when you get a chance. My recommendation is to focus instead on the case where there are no ready threads; this is happen in almost all real applications. (They may have the same or similar solution, though.)

**Comment by newbrain** [comment]
Scratch the quick test part!
It seems to be much better than that: the reported warning was due to using _false_ instead of _False_ (eh, I warned I'm a C programmer...).

The result seems to be quite OK:
```
(gdb) mon reset halt
Resetting target with halt
Successfully halted device on reset
(gdb) tb main
Temporary breakpoint 2 at 0x800041c: file Src/sample_threadx.c, line 62.
(gdb) c
Continuing.

Temporary breakpoint 2, main () at Src/sample_threadx.c:62
62          tx_kernel_enter();
(gdb) c
Continuing.

Program received signal SIGINT, Interrupt.
0x080013dc in _txe_queue_receive (queue_ptr=0x200029b4 <queue_0>, destination_ptr=0x20001138 <memory_area+3032>,
    wait_option=4294967295) at ThreadX/common/src/txe_queue_receive.c:140
140                     if (current_thread == &_tx_timer_thread)
(gdb) info threads
  Id   Target Id         Frame
* 1    Thread 1          0x080013dc in _txe_queue_receive (queue_ptr=0x200029b4 <queue_0>,
    destination_ptr=0x20001138 <memory_area+3032>, wait_option=4294967295)
    at ThreadX/common/src/txe_queue_receive.c:140
(gdb) n
152         if (status == TX_SUCCESS)
(gdb) n
156             status =  _tx_queue_receive(queue_ptr, destination_ptr, wait_option);
(gdb) n
160         return(status);
(gdb) n
161     }
(gdb) n
thread_2_entry (thread_input=2) at Src/sample_threadx.c:248
248             if ((status != TX_SUCCESS) || (received_message != thread_2_messages_received))
(gdb) n
252             thread_2_messages_received++;
(gdb) n
241             thread_2_counter++;
(gdb) info threads
  Id   Target Id         Frame
* 1    Thread 1          thread_2_entry (thread_input=2) at Src/sample_threadx.c:241
(gdb) bt
#0  thread_2_entry (thread_input=2) at Src/sample_threadx.c:241
#1  0x08004392 in _tx_thread_shell_entry () at ThreadX/common/src/tx_thread_shell_entry.c:114
#2  0xfffffffe in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```
But, to make sure, I still will set up a proper test.

**Comment by newbrain** [comment]
The case when no task is Ready is handled by the logic in  [ThreadXThreadContext::read_core_registers_raw](https://github.com/newbrain/pyOCD/blob/f4cb0c8410bf340521b05c6ba026499644891717/pyocd/rtos/threadx.py#L184-L254).
If the thread is not current, the stacked frame will be read, otherwise the live registers.
As this is done thread by thread, there is no need for a special case when all threads are suspended/waiting.

**Comment by flit** [comment]
Perfect! 👍🏽

**Comment by newbrain** [comment]
As a proof, I put all threads to sleep (it's almost midnight here, and they are quite young):
```
 (gdb) info threads
  Id   Target Id                                                      Frame
* 2    Thread 2 "Handler mode" (PendSV)                               __tx_ts_wait ()
    at ThreadX/ports/cortex_m4/gnu/src/tx_thread_schedule.S:251
  3    Thread 536882300 "System Timer Thread" (Suspended; Priority 0) 0x08004916 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  4    Thread 536881408 "thread 0" (Sleep; Priority 1)                0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  5    Thread 536871640 "thread 1" (Sleep; Priority 16)               0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  6    Thread 536871044 "thread 2" (Sleep; Priority 16)               0x08004916 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  7    Thread 536872108 "thread 3" (Sleep; Priority 8)                0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  8    Thread 536871456 "thread 4" (Sleep; Priority 8)                0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  9    Thread 536881748 "thread 5" (Sleep; Priority 4)                0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  10   Thread 536871820 "thread 6" (Sleep; Priority 8)                0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
  11   Thread 536871264 "thread 7" (Sleep; Priority 8)                0x08004a42 in __get_ipsr_value ()
    at ThreadX/ports/cortex_m4/gnu/inc/tx_port.h:371
```

Note the the idle loop in ThreadX is, as you said, inside the scheduler (PendSV ISR), i.e. in the "Handler Mode" thread.
Were it outside, some special handling might have been needed.

**Comment by newbrain** [comment]
After some documentation perusing and test code abusing, the corner case I hypothesized above (scheduler started but zero threads) is not possible: in fact, threads and other objects _cannot_ be created from a timer callback.
If error checking is enabled during compilation, TX_CALLER_ERROR is returned by the likes of tx_*primitive*_create()
So, not something we should worry about.

**Comment by flit** [comment]
All very good, thanks for double checking.

File: pyocd/rtos/threadx.py [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+        if not self.is_enabled:
+            return None
+        self.update_threads()
+        return self._threads.get(threadId, None)
+
+    @property
+    def is_enabled(self):
+        if self._system_state is None:
+            return False
+        return self._target_context.read32(self._system_state) < ThreadXThreadProvider.TX_INITIALIZE_IN_PROGRESS
```
Comment (flit) [inline pyocd/rtos/threadx.py:471]: Is a less-than comparison correct? Just odd to see < a pattern like 0xF0F0F0F0. If it's correct, maybe add comment on how the comparison works?

File: pyocd/rtos/threadx.py [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+                    t = ThreadXThread(self._target_context, self, threadBase)
+                LOG.debug("Thread 0x%08x (%s)", threadBase, t.name)
+                newThreads[t.unique_id] = t
+            except exceptions.TransferError:
+                LOG.debug(
+                    "TransferError while examining thread 0x%08x", threadBase)
+
+        # Is the number of threads correct?
+        if len(newThreads) != threadCount:
+            LOG.warning("ThreadX: thread count mismatch, expected: {0}, found: {1}".format(
```
Comment (flit) [inline pyocd/rtos/threadx.py:444]: Use `%` formatting instead of `format()` method, like the other log calls. The logger stores the string formatting arguments in the log message object and only formats the string if it gets printed.

File: setup.py [hunk 3]
```diff
@@ -123,6 +123,7 @@
             'freertos = pyocd.rtos.freertos:FreeRTOSPlugin',
             'zephyr = pyocd.rtos.zephyr:ZephyrPlugin',
```
Comment (flit) [inline setup.py:124]: Would you mind moving `zephyr` to the end of this list? I should have kept it in sorted order… Thanks.

**Review by flit — CHANGES_REQUESTED** [review]
Looks really good. Just a few small things to change and it's ready to merge.

File: pyocd/rtos/threadx.py [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+        0x4f:   192,    # s15
+        33:     196,    # fpscr
+        # (reserved word: 200)
+    }
+
+    def __init__(self, parent, thread):
+        super(ThreadXThreadContext, self).__init__(parent)
+        self._thread = thread
+        self._has_fpu = self.core.has_fpu
+        self.nofpu_register_offsets = self.NOFPU_REGISTER_OFFSETS
```
[the remark this answers is not in the document]
Comment (newbrain) [reply pyocd/rtos/threadx.py:180]: Good catch, thanks - the case you make is not really far fetched, as e.g. NXP LPC43S37 has two M0 (+?) and one M4 and if a soft abi is used for FP, it will appear just as you envisioned.

File: pyocd/rtos/threadx.py [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+        Semaphore = 6
+        EventFlag = 7
+        BlockMemory = 8
+        ByteMemory = 9
+        IoDriver = 10
+        File = 11
+        TcpIp = 12
+        Mutex = 13
+        PriorityChange = 14
+        Unknown = 99
```
[the remark this answers is not in the document]
Comment (newbrain) [reply pyocd/rtos/threadx.py:276]: I see, on the other hand I was using them also as the printed out name for the thread state (via enum.name).
I will revert to a dictionary (my first implementation, not pushed), if this is against the rules.

File: pyocd/rtos/threadx.py (hunk 1 above)
Comment (newbrain) [reply pyocd/rtos/threadx.py:471]: I will add a comment, the system state takes TX_INITIALIZE_IN_PROGRESS and TX_INITIALIZE_IN_PROGRESS+1 during initialization, and 0 (for cortex-m ports). In other ports, it contains interrupt nesting count.

File: pyocd/rtos/threadx.py (hunk 2 above)
Comment (newbrain) [reply pyocd/rtos/threadx.py:444]: Agreed. Did not know it would make a difference.

File: setup.py (hunk 3 above)
Comment (newbrain) [reply setup.py:124]: Of course! In fact, in my first version I did just that, then thought of not introducing spurious changes.

**Comment by newbrain** [comment]
Thanks for the insightful comments, I'll work on this tonight - hope to have it ready soon.

**Comment by newbrain** [comment]
I don't think I need to rebase on master, as the PR can be merged cleanly.
Let me know in case it's preferred.

**Review by flit — APPROVED** [review]
Perfect!

**Comment by newbrain** [comment]
Thanks for the great support!

**Comment by flit** [comment]
Thanks for the contribution! 😄
